src/tools/ceph_objectstore_tool.cc

```diff
@@ -646,23 +646,33 @@ int do_trim_pg_log(ObjectStore *store, const coll_t &coll,
 
   cerr << "Log bounds are: " << "(" << info.log_tail << ","
        << info.last_update << "]" << std::endl;
-
-  uint64_t max_entries = g_ceph_context->_conf->osd_max_pg_log_entries;
+  uint64_t max_entries = std::min<uint64_t>(g_ceph_context->_conf->osd_max_pg_log_entries,
+					    (info.last_update.version) ? (info.last_update.version - 1)
+					        : g_ceph_context->_conf->osd_max_pg_log_entries
+					    );
   if (info.last_update.version - info.log_tail.version <= max_entries) {
     cerr << "Log not larger than osd_max_pg_log_entries " << max_entries << std::endl;
     return 0;
   }
 
+  PGLog::IndexedLog pg_log;
+  pg_log.head = info.last_update;
+  pg_log.skip_can_rollback_to_to_head();
+
   ceph_assert(info.last_update.version > max_entries);
-  version_t trim_to = info.last_update.version - max_entries;
+  eversion_t trim_to(info.last_update.epoch, info.last_update.version - max_entries);
```

Comment:
using eversion_t like this will trim many more entries since it's sorted by epoch, then version. don't think it  matters in practice in this case since this is an emergency recovery tool though.